internal/ctrlutils/ingress-status.go

```diff
@@ -41,35 +41,31 @@ func PullConfigUpdate(
 	kubeConfig *rest.Config,
 	publishService string,
 	publishAddresses []string,
-) {
+) <-chan error {
```

Comment:
This signature seems problematic: this seems to indicate that the caller should expect a potentially limitless number of errors to be produced by this component and add caller-level logic to handle each of those (sounds like for now it will probably just log those).

It seems like it might be better to simply return an `error` here if anything fails in the setup steps of this, and letting the logger log runtime errors for the persistent goroutine for non-fatal error conditions. I recommend we re-work this so that several of the following conditions (like not being able to determine the ips/hostnames of the kong proxy) simply return an error and indicate that this server was unable to start properly.

Reply:
This error channel populates update status thread error into main thread, and we log this error since already have reasonable retry in the update mechanism. Channel is introduced because this is cross-goroutine communication, pure return value could not resolve the issue.

Reply:
> Channel is introduced because this is cross-goroutine communication, pure return value could not resolve the issue.

I think there may be a misunderstanding here because I'm really confused by this comment.

Perhaps a clarifying question would be helpful:

Q: what do you expect the caller should be doing with the `chan error` they receive after calling this function?

Reply:
The logged error is useful to user:  they can manually check the ingress status, configuration if an error happened.  
Also, this is part of KIC metrics.

Reply:
Ok, so it's just for logging. In that case I'm still thinking the original comment has value here: fail on all the setup stuff, but log the runtime errors that are in the server over time. otherwise the calling code will need to make a separate goroutine just to log errors for this one server.

Reply:
Please be noted the diff is a channel is used to communicate errors between update and main thread, the update goroutine has been applied and function well
https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/manager/run.go#L109
The channel helps to populate error, which pure return could not achive.

Reply:
Offline discussion, we agreed on 
- keep the error channel between go-routines
- create an issue - report update status metrics using controller-runtime
- exit the thread when setup failed https://github.com/Kong/kubernetes-ingress-controller/pull/1758/files#diff-acd60b643e0d662a409aa00230a4a34bb0ae4dee18903d773d757277edc26340L46-R68

Reply:
5ac1d90c5e6aa8319c4fd3169d6e45da4810e7cf

Reply:
issue created https://github.com/Kong/kubernetes-ingress-controller/issues/1847